compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala

```diff
@@ -37,6 +38,14 @@ object ProtoTypes {
     def isCompatible(tp: Type, pt: Type)(using Context): Boolean =
       (tp.widenExpr relaxed_<:< pt.widenExpr) || viewExists(tp, pt)
 
+    /** Like isCompatibe, but using a subtype comparison with necessary eithers
+     *  that don't unnecessarily truncate the constraint space, returning false instead.
+     */
+    def necessarilyCompatible(tp: Type, pt: Type)(using Context): Boolean =
+      val tpw = tp.widenExpr
+      val ptw = pt.widenExpr
+      necessarySubType(tpw, ptw) || tpw.isValueSubType(ptw) || viewExists(tp, pt)
```

Comment:
Does it matter?

```suggestion
      necessarySubType(tpw, ptw) || tpw.isValueSubType(ptw) || viewExists(tpw, ptw)
```

Reply:
I think it would matter if there is an implicit conversion between singleton types. In this case it should be `viewExists(tp, pt)`